File: klinara-ios/klinara-ios/Services/Formatting/Money.swift
import Foundation

/// Para biçimlendirme ve ayrıştırma — **tek yer**.
///
/// Sunucu tarafındaki kural burada da geçerlidir: tutarlar kuruş cinsinden
/// `Int` taşınır, `Double`'a **çevrilmez**. Bir kez `Double`'a düşen tutar
/// 0,1 + 0,2 ≠ 0,3 dünyasına girer ve taksit toplamları tutmamaya başlar.
enum Money {

    nonisolated(unsafe) private static let formatter: NumberFormatter = {
        let formatter = NumberFormatter()
        formatter.numberStyle = .currency
        formatter.locale = Locale(identifier: "tr_TR")
        formatter.currencyCode = "TRY"
        formatter.minimumFractionDigits = 2
        formatter.maximumFractionDigits = 2
        return formatter
    }()

    /// Kuruş → "1.500,00 ₺".
    static func format(minor: Int, currency: String = "TRY") -> String {
        // `NSDecimalNumber` ile bölüyoruz: 250_000 / 100 tamsayı bölmesinde
        // kuruşu düşürürdü, `Double` bölmesi ise 2.500,0000000001 üretebilirdi.
        let value = NSDecimalNumber(value: minor)
            .dividing(by: 100, withBehavior: roundingBehavior)
        formatter.currencyCode = currency
        return formatter.string(from: value) ?? "\(minor)"
    }

    /// Sembolsüz biçim — giriş alanının içinde sembol tekrar edilmesin diye.
    static func formatPlain(minor: Int) -> String {
        let value = NSDecimalNumber(value: minor).dividing(by: 100, withBehavior: roundingBehavior)
        return plainFormatter.string(from: value) ?? "\(minor)"
    }

    /// Kullanıcının yazdığı metni kuruşa çevirir. "1.500,50" ve "1500.50" ikisi de kabul.
    /// Ayrıştırılamayan girdi `nil` döner — sessizce 0 kabul etmek, kullanıcının
    /// yazdığından farklı bir fiyat kaydetmenin en kestirme yoludur.
    static func parse(_ text: String) -> Int? {
        let stripped = text
            .replacingOccurrences(of: "₺", with: "")
            .replacingOccurrences(of: " ", with: "")
            .replacingOccurrences(of: "\u{00A0}", with: "")
            .trimmingCharacters(in: .whitespaces)

        let cleaned = normalizeSeparators(stripped)
        guard !cleaned.isEmpty else { return nil }
        let decimal = NSDecimalNumber(string: cleaned, locale: Locale(identifier: "en_US_POSIX"))
        guard decimal != .notANumber else { return nil }

        let minor = decimal.multiplying(by: 100, withBehavior: roundingBehavior)
        guard minor.compare(NSDecimalNumber.zero) != .orderedAscending else { return nil }
        return minor.intValue
    }

    /// Nokta ve virgülün hangisinin ondalık olduğunu ayırır.
    ///
    /// Virgül varsa Türkçe yazım kesindir: noktalar binlik ayırıcıdır. Virgül
    /// yoksa tek bir noktanın ardından **bir ya da iki** basamak geliyorsa bu
    /// ondalıktır (`1500.50`), üç basamak geliyorsa binlik ayırıcıdır (`1.500`).
    ///
    /// Bu ayrım olmadan `"1500.50"` girişi noktası silinip `150050` oluyor ve
    /// 1.500,50 ₺ yerine **150.050,00 ₺** kaydediliyordu — yüz katlık bir hata.
    private static func normalizeSeparators(_ raw: String) -> String {
        if raw.contains(",") {
            return raw
                .replacingOccurrences(of: ".", with: "")
                .replacingOccurrences(of: ",", with: ".")
        }
        let parts = raw.components(separatedBy: ".")
        if parts.count == 2, (1...2).contains(parts[1].count) {
            return raw
        }
        return raw.replacingOccurrences(of: ".", with: "")
    }

    private static let roundingBehavior = NSDecimalNumberHandler(
        roundingMode: .bankers,
        scale: 2,
        raiseOnExactness: false,
        raiseOnOverflow: false,
        raiseOnUnderflow: false,
        raiseOnDivideByZero: false
    )

    nonisolated(unsafe) private static let plainFormatter: NumberFormatter = {
        let formatter = NumberFormatter()
        formatter.numberStyle = .decimal
        formatter.locale = Locale(identifier: "tr_TR")
        formatter.minimumFractionDigits = 2
        formatter.maximumFractionDigits = 2
        return formatter
    }()
}

/// KDV oranı — sunucu **baz puan** (bps) taşır: %20 = 2000.
enum VatRate {

    static func format(basisPoints: Int) -> String {
        let percent = Double(basisPoints) / 100
        return percent == percent.rounded()
            ? "%\(Int(percent))"
            : String(format: "%%%.2f", percent)
    }

    /// Yaygın Türkiye KDV oranları — seçici için.
    static let common: [Int] = [0, 1000, 2000]
}

/// Süre biçimlendirme: 90 → "1 sa 30 dk".
enum DurationFormat {

    static func format(minutes: Int) -> String {
        let hours = minutes / 60
        let remaining = minutes % 60
        return switch (hours, remaining) {
        case (0, _): "\(remaining) dk"
        case (_, 0): "\(hours) sa"
        default: "\(hours) sa \(remaining) dk"
        }
    }
}

File: klinara-android/app/src/main/java/com/klinara/android/services/formatting/Money.kt
package com.klinara.android.services.formatting

import java.math.BigDecimal
import java.math.RoundingMode
import java.text.DecimalFormat
import java.text.DecimalFormatSymbols
import java.util.Locale

/** Türkçe biçimlendirme yereli. */
internal val TrLocale: Locale = Locale.forLanguageTag("tr-TR")

/**
 * Para biçimlendirme ve ayrıştırma — **tek yer**.
 *
 * Sunucu tutarları `bigint` **kuruş** olarak gönderiyor; istemcide `Long` kalır.
 * `Double`/`Float` YASAK: bir kez kayan noktaya düşen tutar 0,1 + 0,2 ≠ 0,3 dünyasına
 * girer ve ciro toplamlarında kuruş hatası birikir.
 *
 * **`NumberFormat.getCurrencyInstance` KULLANILMAZ.** JVM (JDK CLDR) ve Android (ICU)
 * aynı yerel için farklı çıktı verir — sembol konumu ve NBSP vs normal boşluk. Birim
 * testleri yeşil olur, cihazda çıktı yanlış olurdu. Ayırıcılar ve sembol AÇIKÇA
 * kuruluyor; böylece JVM ile cihaz yapı gereği aynı.
 */
object Money {
    private const val MINOR_UNITS_PER_MAJOR = 100L
    private const val FRACTION_DIGITS = 2
    private const val LIRA_SYMBOL = "₺"

    private val symbols =
        DecimalFormatSymbols(TrLocale).apply {
            groupingSeparator = '.'
            decimalSeparator = ','
        }

    private fun formatter() =
        DecimalFormat("#,##0.00", symbols).apply {
            roundingMode = RoundingMode.HALF_EVEN
        }

    /** Kuruş → "1.500,00 ₺". */
    fun format(
        minor: Long,
        currency: String = "TRY",
    ): String {
        val amount = formatPlain(minor)
        val symbol = if (currency == "TRY") LIRA_SYMBOL else currency
        return "$amount $symbol"
    }

    /** Sembolsüz biçim — giriş alanının içinde sembol tekrar edilmesin diye. */
    fun formatPlain(minor: Long): String =
        formatter().format(
            BigDecimal.valueOf(minor).divide(
                BigDecimal.valueOf(MINOR_UNITS_PER_MAJOR),
                FRACTION_DIGITS,
                // Bankacı yuvarlaması: iOS `NSDecimalNumberHandler(.bankers)` paritesi.
                RoundingMode.HALF_EVEN,
            ),
        )

    /**
     * Kullanıcının yazdığı metni kuruşa çevirir. "1.500,50" ve "1500.50" ikisi de kabul.
     *
     * Ayrıştırılamayan girdi `null` döner — sessizce 0 kabul etmek, kullanıcının
     * yazdığından farklı bir fiyat kaydetmenin en kestirme yoludur.
     */
    fun parse(text: String): Long? {
        val stripped =
            text
                .replace(LIRA_SYMBOL, "")
                .replace(" ", "")
                .replace(" ", "")
                .trim()

        val cleaned = normalizeSeparators(stripped)
        if (cleaned.isEmpty()) return null

        val decimal = runCatching { BigDecimal(cleaned) }.getOrNull() ?: return null
        if (decimal.signum() < 0) return null

        return decimal
            .multiply(BigDecimal.valueOf(MINOR_UNITS_PER_MAJOR))
            .setScale(0, RoundingMode.HALF_EVEN)
            .toLong()
    }

    /**
     * Nokta ve virgülün hangisinin ondalık olduğunu ayırır.
     *
     * Virgül varsa Türkçe yazım kesindir: noktalar binlik ayırıcıdır. Virgül yoksa tek
     * bir noktanın ardından **bir ya da iki** basamak geliyorsa bu ondalıktır
     * (`1500.50`), üç basamak geliyorsa binlik ayırıcıdır (`1.500`).
     *
     * Bu ayrım olmadan `"1500.50"` girişi noktası silinip `150050` oluyor ve 1.500,50 ₺
     * yerine **150.050,00 ₺** kaydediliyordu — yüz katlık bir hata.
     */
    private fun normalizeSeparators(raw: String): String {
        if (raw.contains(',')) {
            return raw.replace(".", "").replace(',', '.')
        }
        val parts = raw.split('.')
        if (parts.size == 2 && parts[1].length in 1..FRACTION_DIGITS) return raw
        return raw.replace(".", "")
    }
}

/** KDV oranı — sunucu **baz puan** (bps) taşır: %20 = 2000. */
object VatRate {
    private const val BPS_PER_PERCENT = 100

    fun format(basisPoints: Int): String {
        val whole = basisPoints / BPS_PER_PERCENT
        val remainder = basisPoints % BPS_PER_PERCENT
        return if (remainder == 0) {
            "%$whole"
        } else {
            "%" + Money.formatPlain(basisPoints.toLong())
        }
    }

    const val ZERO = 0
    const val TEN_PERCENT = 1000
    const val TWENTY_PERCENT = 2000

    /** Yaygın Türkiye KDV oranları — seçici için. */
    val common: List<Int> = listOf(ZERO, TEN_PERCENT, TWENTY_PERCENT)
}

/** Süre biçimlendirme: 90 → "1 sa 30 dk". */
object DurationFormat {
    private const val MINUTES_PER_HOUR = 60

    fun format(minutes: Int): String {
        val hours = minutes / MINUTES_PER_HOUR
        val remaining = minutes % MINUTES_PER_HOUR
        return when {
            hours == 0 -> "$remaining dk"
            remaining == 0 -> "$hours sa"
            else -> "$hours sa $remaining dk"
        }
    }
}
